grin/src/Test/ExtendedSyntax/New/Test.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE QuasiQuotes, ViewPatterns #-}
+module Test.ExtendedSyntax.New.Test where
+
+import Data.Text (Text, pack)
+import Data.Bifunctor (second)
+
+import Control.Monad (forM_, forM)
+
+import Test.Hspec (Spec, describe)
+
+import Grin.ExtendedSyntax.TH (expr)
+import Grin.ExtendedSyntax.Grin (Exp, Name)
+import Grin.ExtendedSyntax.TypeEnv (TypeEnv, emptyTypeEnv)
+import Grin.ExtendedSyntax.Pretty (Pretty, PP(..))
+
+import Transformations.ExtendedSyntax.Names (evalNameM, deriveNewName)
+
+type SpecWithProg = Exp -> Spec
+
+type TestExpContext = (String, (TypeEnv, Exp) -> (TypeEnv, Exp))
+
+testExprContext :: (((TypeEnv, Exp) -> (TypeEnv, Exp)) -> Spec) -> Spec
+testExprContext mkSpec = forM_ contexts $ \(label, ctx) -> describe (concat ["(", label, ")"]) $ mkSpec ctx
+
+testExprContextE :: ((Exp -> Exp) -> Spec) -> Spec
+testExprContextE mkSpec =
+  forM_ contexts $ \(label, ctx) ->
+    describe (concat ["(", label, ")"]) $ mkSpec (\e -> snd $ ctx (emptyTypeEnv, e))
+
+contexts :: [TestExpContext]
+contexts =
+  [ emptyCtx
+  , lastBindR
+  , bindL
+  , lastBindL
+  , firstAlt
+  , middleAlt
+  , lastAlt
+  ]
+
+emptyCtx :: TestExpContext
+emptyCtx = ("empty", id)
+
+-- NOTE: These contexts contain some names. Make sure not to use these in your test code!
+
+deriveNames :: Exp -> [Name]
+deriveNames e = fst <$> evalNameM e $ do
+  forM [1..] $ \_ -> deriveNewName "ctxVar"
+
+deriveNamesAsText :: Exp -> [Text]
+deriveNamesAsText = map toText . deriveNames
+
+toText :: Pretty a => a -> Text
+toText = pack . show . PP
+
+bindL :: TestExpContext
+bindL = ("bind left", second tr) where
+  tr e
+    | (v:_) <- deriveNamesAsText e
+    , eText <- toText e = [expr|
+      $v <- do
+        $eText
+      pure ()
+    |]
+
+lastBindL :: TestExpContext
+lastBindL = ("last bind left", second tr) where
+  tr e
+    | (v1:v2:_) <- deriveNamesAsText e
+    , eText <- toText e = [expr|
+      $v1 <- do
```

Comment:
Does the '$v1' here mean that we generate new names in this context based on the expression 'e'.

Reply:
Yes. The generated context will contain fresh names.